backend/src/main/kotlin/app/ehrenamtskarte/backend/verification/Argon2IdHasher.kt

```diff
@@ -0,0 +1,73 @@
+import app.ehrenamtskarte.backend.user.KoblenzUser
+import app.ehrenamtskarte.backend.verification.CanonicalJson
+import org.bouncycastle.crypto.generators.Argon2BytesGenerator
+import org.bouncycastle.crypto.params.Argon2Parameters
+import java.nio.charset.StandardCharsets
+import java.util.Base64
+
+class Argon2IdHasher {
+    companion object {
+        /**
+         * Copied from spring-security Argon2EncodingUtils.java licenced under Apache 2.0, removed the salt from the result
+         *
+         * Encodes a raw Argon2-hash and its parameters into the standard Argon2-hash-string
+         * as specified in the reference implementation
+         * (https://github.com/P-H-C/phc-winner-argon2/blob/master/src/encoding.c#L244):
+         *
+         * {@code $argon2<T>[$v=<num>]$m=<num>,t=<num>,p=<num>$<bin>$<bin>}
+         **/
+        @Throws(IllegalArgumentException::class)
+        fun encode(
+            hash: ByteArray?,
+            parameters: Argon2Parameters
+        ): String? {
+            val b64encoder: Base64.Encoder = Base64.getEncoder().withoutPadding()
+            val stringBuilder = StringBuilder()
+            val type =
+                when (parameters.type) {
+                    Argon2Parameters.ARGON2_d -> "\$argon2d"
+                    Argon2Parameters.ARGON2_i -> "\$argon2i"
+                    Argon2Parameters.ARGON2_id -> "\$argon2id"
+                    else -> throw IllegalArgumentException("Invalid algorithm type: " + parameters.type)
+                }
+            stringBuilder.append(type)
+            stringBuilder
+                .append("\$v=")
+                .append(parameters.version)
+                .append("\$m=")
+                .append(parameters.memory)
+                .append(",t=")
+                .append(parameters.iterations)
+                .append(",p=")
+                .append(parameters.lanes)
+            stringBuilder.append("$").append(b64encoder.encodeToString(hash))
+            return stringBuilder.toString()
+        }
+
+        fun hashKoblenzUserData(userData: KoblenzUser): String? {
+            val canonicalJson = CanonicalJson.koblenzUserToString(userData)
+            val hashLength = 32
+
+            val pepper = Environment.getVariable(KOBLENZ_PEPPER_SYS_ENV) // TODO handle if Null
+            val pepperByteArray = pepper?.toByteArray(StandardCharsets.UTF_8)
+            val params =
+                Argon2Parameters
+                    .Builder(Argon2Parameters.ARGON2_id)
+                    .withVersion(19)
+                    .withIterations(2)
+                    .withSalt(pepperByteArray)
+                    .withParallelism(1)
+                    .withMemoryAsKB(16)
+                    .build()
+
+            val generator = Argon2BytesGenerator()
+            generator.init(params)
+            val result = ByteArray(hashLength)
+            generator.generateBytes(canonicalJson.toByteArray(), result)
```

Comment:
This comment is more related to CanonicalJson. I just briefly checked its implementation and it seems like it is not enforing an order of the fields. It seems like the protobuf implementation determines how things are sorted: https://github.com/digitalfabrik/entitlementcard/blob/aa83fa44fede13446b1f1b48f3886f4f525dd754/backend/src/main/kotlin/app/ehrenamtskarte/backend/cards/CanonicalJson.kt#L61-L65


I could be wrong here though. If the above is true we should probably test the CanonicalJson implementation.

Reply:
How does the canonical json react to inputs that need to be escaped? E.g. names that include a quote?

Reply:
The order is canonicalized here: https://github.com/digitalfabrik/entitlementcard/pull/1499/files/eb2d7bbd19f7ee123a4725e74ba8bb473252022f#diff-c688a2cede2d0e62a8d75f41502f13cf619cb1d3d5c126b7ccc75329c3da8fc1R135

The escaping is (probably/hopefully) done by https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-json/kotlinx.serialization.json/-json-primitive/ (which we construct here https://github.com/digitalfabrik/entitlementcard/pull/1499/files/eb2d7bbd19f7ee123a4725e74ba8bb473252022f#diff-c688a2cede2d0e62a8d75f41502f13cf619cb1d3d5c126b7ccc75329c3da8fc1R125 )